plutus-core/testlib/PlutusCore/Generators/QuickCheck/Utils.hs

```diff
@@ -218,3 +220,90 @@ multiSplit xs = do
+            then ([y | keep] ++) <$> go keep weight xs
+            else case xs of
+                []      -> pure []
+                x : xs' -> (x :) <$> go True (weight * scaling) xs'
+
+-- | Insert a value into a list an arbitrary number of times. The first argument controls whether
+-- to allow inserting at the end of the list, the second argument is the probability of
+-- inserting an element at the beginning of the list.
+insertManyPreferLeft :: Bool -> Double -> a -> [a] -> Gen [a]
+insertManyPreferLeft keepSuffix headProb y =
+    fmap reverse . insertManyPreferRight keepSuffix headProb y . reverse
+
+-- | Insert a value into a list an arbitrary number of times. The first argument is the probability
+-- of inserting an element at an end of the list (i.e. either the beginning or the end, not
+-- combined). See 'multiSplit' for what this function allows us to do.
+insertManyPreferEnds :: Double -> a -> [a] -> Gen [a]
+-- Cut the list in half, insert into the left half skewing generation towards the beginning, insert
+-- into the right half skewing generation towards the end, then append the results of those two
+-- operations, so that we get a list where additional elements are more likely to occur close to
+-- the sides.
+insertManyPreferEnds endProb y xs = do
+    -- In order not to get skewed results we sometimes put the middle element of the list into its
+    -- first half and sometimes into its second half.
+    isOddToLeft <- arbitrary
+    let (xsL, xsR) = halve isOddToLeft xs
+    -- If the list has even length, then it was cut into two halves of equal length meaning one slot
+    -- for to put an element in appears twice: at the end of the left half and at the beginning of
+    -- the right one. Hence in order to avoid skeweness we don't put anything into this slot at the
+    -- end of the left half.
+    -- Maybe we do want to skew generation to favor the middle of the list like we do for its ends,
+    -- but then we need to do that intentionally and systematically, not randomly and a little bit.
+    xsL' <- insertManyPreferLeft (length xsL /= length xsR) endProb y xsL
+    xsR' <- insertManyPreferRight True endProb y xsR
+    pure $ xsL' ++ xsR'
+
+-- | Split a list into chunks at random. Concatenating the resulting lists gives back the original
+-- one. Generates empty chunks. The first argument is the probability of generating at least one
+-- empty chunk as the first element of the resulting list. It is also the probability of generating
+-- an empty chunk as the last element of the resulting list. The probability of generating empty
+-- chunks decreases as we go from either of the ends of the resulting list (this is so that we are
+-- more likely to hit a corner case related to handling elements at the beginning or the end of a
+-- list).
+multiSplit :: Double -> [a] -> Gen [[a]]
+multiSplit endProb = multiSplit1 >=> insertManyPreferEnds endProb [] . coerce
```

Comment:
The tests are more complicated than the code they're testing now!  I'm reasonably convinced (mostly by looking at the output of the generators) that they're doing the right things though.

Reply:
Oh, I forgot to report here. I did look at the output of the generators closely for much larger probability of zeroes and more test runs and it was quite great (newlines added for readability):

```
          +++ OK, passed 100000 tests:
          12.967% [1]

           9.641% [0,1]
           9.614% [1,0]

           6.346% [1,0,0]
           6.341% [0,0,1]
           6.305% [0,1,0]

           4.087% [0,1,0,0]
           3.898% [1,0,0,0]
           3.875% [0,0,1,0]
           3.850% [0,0,0,1]

           2.367% [0,1,0,0,0]
           2.344% [1,0,0,0,0]
           2.336% [0,0,0,0,1]
           2.308% [0,0,0,1,0]
           2.293% [0,0,1,0,0]

           1.384% [0,0,0,1,0,0]
           1.377% [0,1,0,0,0,0]
           1.360% [0,0,1,0,0,0]
           1.314% [0,0,0,0,1,0]
           1.295% [1,0,0,0,0,0]
           1.271% [0,0,0,0,0,1]

           0.809% [0,0,0,1,0,0,0]
           0.759% [0,0,1,0,0,0,0]
           0.758% [0,0,0,0,1,0,0]
           0.753% [0,1,0,0,0,0,0]
           0.745% [0,0,0,0,0,0,1]
           0.720% [1,0,0,0,0,0,0]
           0.718% [0,0,0,0,0,1,0]
           <...>
```

and

```
           6.538% [1,1]
           6.390% [2]

           4.830% [0,2]
           4.828% [2,0]

           3.326% [0,1,1]
           3.269% [0,2,0]
           3.221% [1,1,0]
           3.146% [1,0,1]
           3.140% [0,0,2]
           3.133% [2,0,0]

           1.969% [2,0,0,0]
           1.949% [0,2,0,0]
           1.918% [0,0,2,0]
           1.901% [0,0,0,2]

           1.658% [1,1,0,0]
           1.634% [0,1,1,0]
           1.582% [0,0,1,1]
           1.559% [0,1,0,1]
           1.508% [1,0,1,0]
           1.475% [1,0,0,1]

           1.188% [0,0,2,0,0]
           1.158% [2,0,0,0,0]
           1.147% [0,0,0,0,2]
           1.138% [0,2,0,0,0]
           1.119% [0,0,0,2,0]

           0.874% [0,0,0,1,1]
           0.859% [1,1,0,0,0]
           0.823% [0,0,1,0,1]
           0.817% [0,0,1,1,0]
           0.807% [0,1,1,0,0]
           0.771% [1,0,1,0,0]
           0.741% [0,1,0,0,1]
           0.733% [0,1,0,1,0]
           0.732% [1,0,0,1,0]
```

You can see that there's some skewness in for example

```
           1.658% [1,1,0,0]
           1.634% [0,1,1,0]
           1.582% [0,0,1,1]
           1.559% [0,1,0,1]
           1.508% [1,0,1,0]
           1.475% [1,0,0,1]
```

but that's by design, since I made it so that inserting a zero (or an empty list etc) in the middle of a list happens a bit less often than doing it at one of the ends, which is precisely what we're observing.